File: Practica_3_java/src/Logica/Cola.java

package Logica;


public class Cola {
    private long[] arreglo;  // Arreglo que almacenará los elementos de la cola
    private int inicio;      // Índice que indica el primer elemento de la cola
    private int fin;         // Índice que indica el último elemento de la cola
    private int n;           // Tamaño máximo de la cola

    // Constructor de la clase Cola
    public Cola(int n) {
        this.n = n;
        this.arreglo = new long[n]; // Inicializa el arreglo con el tamaño dado
        this.inicio = -1;  // La cola está vacía inicialmente, no hay elementos
        this.fin = -1;     // La cola está vacía inicialmente
    }

    // Método para insertar un elemento en la cola
    public void insert(long e) {
        if (isFull()) {
            System.out.println("La cola está llena. No se puede insertar el elemento.");
        } else {
            if (inicio == -1) {
                inicio = 0; // Si la cola está vacía, el primer elemento se coloca en el índice 0
            }
            fin = (fin + 1) % n;  // Calcula la siguiente posición de fin (uso de módulo para el comportamiento circular)
            arreglo[fin] = e;  // Inserta el elemento en la posición de fin
        }
    }

    // Método para eliminar y devolver el primer elemento de la cola
    public long remove() {
        if (isEmpty()) {
            System.out.println("La cola está vacía. No se puede eliminar un elemento.");
            return -1;  // Valor de error, puedes cambiarlo si lo deseas
        } else {
            long elemento = arreglo[inicio];  // Guarda el primer elemento
            if (inicio == fin) {
                inicio = fin = -1;  // La cola queda vacía
            } else {
                inicio = (inicio + 1) % n;  // Calcula la siguiente posición de inicio
            }
            return elemento;  // Retorna el primer elemento
        }
    }

    // Método para obtener el primer elemento sin eliminarlo
    public long peek() {
        if (isEmpty()) {
            System.out.println("La cola está vacía. No se puede ver el primer elemento.");
            return -1;  // Valor de error, puedes cambiarlo si lo deseas
        } else {
            return arreglo[inicio];  // Retorna el primer elemento sin modificar la cola
        }
    }

    // Método para verificar si la cola está vacía
    public boolean isEmpty() {
        return inicio == -1;  // Si el índice de inicio es -1, la cola está vacía
    }

    // Método para verificar si la cola está llena
    public boolean isFull() {
        return (fin + 1) % n == inicio;  // Si la siguiente posición de fin es igual a inicio, la cola está llena
    }

    // Método para obtener el tamaño de la cola
    public int size() {
        if (isEmpty()) {
            return 0;  // Si la cola está vacía, el tamaño es 0
        } else {
            if (fin >= inicio) {
                return fin - inicio + 1;  // La cola no está circular
            } else {
                return n - inicio + fin + 1;  // La cola es circular
            }
        }
    }
}

File: Practica_3_python/Cola.py
class Cola:
    def __init__(self, n):
        self.n = n  # Tamaño máximo de la cola
        self.arreglo = [0] * n  # Inicializa el arreglo con el tamaño dado
        self.inicio = -1  # El primer índice de la cola
        self.fin = -1  # El último índice de la cola

    # Método para insertar un elemento en la cola
    def insert(self, e):
        if self.isFull():
            print("La cola está llena. No se puede insertar el elemento.")
        else:
            if self.inicio == -1:  # Si la cola está vacía
                self.inicio = 0
            self.fin = (self.fin + 1) % self.n  # Comportamiento circular
            self.arreglo[self.fin] = e

    # Método para eliminar y devolver el primer elemento de la cola
    def long_remove(self):
        if self.isEmpty():
            print("La cola está vacía. No se puede eliminar un elemento.")
            return -1  # Valor de error
        else:
            valor = self.arreglo[self.inicio]
            if self.inicio == self.fin:  # Si hay un solo elemento
                self.inicio = self.fin = -1  # La cola queda vacía
            else:
                self.inicio = (self.inicio + 1) % self.n  # Comportamiento circular
            return valor

    # Método para obtener el primer elemento de la cola sin eliminarlo
    def long_peek(self):
        if self.isEmpty():
            print("La cola está vacía. No se puede ver el primer elemento.")
            return -1  # Valor de error
        else:
            return self.arreglo[self.inicio]

    # Método para verificar si la cola está vacía
    def isEmpty(self):
        return self.inicio == -1  # Si inicio es -1, la cola está vacía

    # Método para verificar si la cola está llena
    def isFull(self):
        return (self.fin + 1) % self.n == self.inicio  # Si la siguiente posición de fin es igual a inicio, la cola está llena

    # Método para obtener el tamaño de la cola
    def size(self):
        if self.isEmpty():
            return 0  # Si la cola está vacía, el tamaño es 0
        elif self.fin >= self.inicio:
            return self.fin - self.inicio + 1  # Si no hay envolvimiento circular
        else:
            return self.n - self.inicio + self.fin + 1  # Si la cola está circular


# Crear una cola con un tamaño máximo de 5
mi_cola = Cola(5)

# Insertar elementos en la cola
mi_cola.insert(1)
mi_cola.insert(2)
mi_cola.insert(3)
mi_cola.insert(4)
mi_cola.insert(5)
# Mostrar el primer elemento
print("Elemento al frente:", mi_cola.long_peek())

# Eliminar y mostrar el primer elemento
print("Elemento eliminado:", mi_cola.long_remove())

# Llenar la cola
mi_cola.insert(1)

# Mostrar el nuevo primer elemento después de hacer pop
print("Nuevo elemento al frente después de remove:", mi_cola.long_peek())

# Verificar si la cola está vacía o llena
print("¿Está vacía la cola?", mi_cola.isEmpty())
print("¿Está llena la cola?", mi_cola.isFull())

# Mostrar el tamaño de la cola
print("Tamaño de la cola:", mi_cola.size())
